File: agent-go/examples/server/agent.go
package main

import (
	"context"
	"fmt"
	"strings"
	"time"

	llmagent "github.com/hoangvvo/llm-sdk/agent-go"
	llmmcp "github.com/hoangvvo/llm-sdk/agent-go/mcp"
	llmsdk "github.com/hoangvvo/llm-sdk/sdk-go"
)

var instructions = []llmagent.InstructionParam[*MyContext]{
	{String: ptr(`Answer in markdown format.
To access certain tools, the user may have to provide corresponding API keys in the context fields on the UI.`)},
	{Func: func(ctx context.Context, context *MyContext) (string, error) {
		name := "<not provided>"
		if context.Name != nil {
			name = *context.Name
		}
		location := "<not provided>"
		if context.Location != nil {
			location = *context.Location
		}
		language := "<not provided>"
		if context.Language != nil {
			language = *context.Language
		}
		return fmt.Sprintf(`The user name is %s.
The user location is %s.
The user speaks %s language.`, name, location, language), nil
	}},
	{Func: func(ctx context.Context, context *MyContext) (string, error) {
		return fmt.Sprintf("The current date is %s.", time.Now().Format("Mon Jan 2 2006")), nil
	}},
	{String: ptr(`For substantive deliverables (documents/specs/code), use the artifact tools (artifact_create, artifact_update, artifact_get, artifact_list, artifact_delete).
Keep chat replies brief and put the full document content into artifacts via these tools, rather than pasting large content into chat. Reference documents by their id.`)},
}

var availableTools = []llmagent.AgentTool[*MyContext]{
	&ArtifactCreateTool{},
	&ArtifactUpdateTool{},
	&ArtifactGetTool{},
	&ArtifactListTool{},
	&ArtifactDeleteTool{},
	&GetStockPriceTool{},
	&GetCryptoPriceTool{},
	&SearchWikipediaTool{},
	&GetNewsTool{},
	&GetCoordinatesTool{},
	&GetWeatherTool{},
}

type AgentOptions struct {
	EnabledTools     []string
	MCPServers       []llmmcp.MCPParams
	Temperature      *float64
	TopP             *float64
	TopK             *int
	FrequencyPenalty *float64
	PresencePenalty  *float64
	Audio            *llmsdk.AudioOptions
	Reasoning        *llmsdk.ReasoningOptions
	Modalities       []llmsdk.Modality
}

func createAgent(model llmsdk.LanguageModel, options *AgentOptions) *llmagent.Agent[*MyContext] {
	var tools []llmagent.AgentTool[*MyContext]
	if options.EnabledTools != nil {
		toolNameSet := make(map[string]bool)
		for _, name := range options.EnabledTools {
			toolNameSet[name] = true
		}
		for _, tool := range availableTools {
			if toolNameSet[tool.Name()] {
				tools = append(tools, tool)
			}
		}
	} else {
		tools = availableTools
	}

	opts := []llmagent.AgentParamsOption[*MyContext]{
		llmagent.WithInstructions(instructions...),
		llmagent.WithTools(tools...),
		llmagent.WithMaxTurns[*MyContext](5),
	}

	if mcpToolkits := createMcpToolkits(options.MCPServers); len(mcpToolkits) > 0 {
		opts = append(opts, llmagent.WithToolkits(mcpToolkits...))
	}

	if options.Temperature != nil {
		opts = append(opts, llmagent.WithTemperature[*MyContext](*options.Temperature))
	}
	if options.TopP != nil {
		opts = append(opts, llmagent.WithTopP[*MyContext](*options.TopP))
	}
	if options.TopK != nil {
		opts = append(opts, llmagent.WithTopK[*MyContext](int32(*options.TopK)))
	}
	if options.FrequencyPenalty != nil {
		opts = append(opts, llmagent.WithFrequencyPenalty[*MyContext](*options.FrequencyPenalty))
	}
	if options.PresencePenalty != nil {
		opts = append(opts, llmagent.WithPresencePenalty[*MyContext](*options.PresencePenalty))
	}
	if options.Reasoning != nil {
		opts = append(opts, llmagent.WithReasoning[*MyContext](*options.Reasoning))
	}
	if options.Audio != nil {
		opts = append(opts, llmagent.WithAudio[*MyContext](*options.Audio))
	}
	if options.Modalities != nil {
		opts = append(opts, llmagent.WithModalities[*MyContext](options.Modalities...))
	}

	return llmagent.NewAgent("MyAgent", model, opts...)
}

func createMcpToolkits(servers []llmmcp.MCPParams) []llmagent.Toolkit[*MyContext] {
	if len(servers) == 0 {
		return nil
	}

	toolkits := make([]llmagent.Toolkit[*MyContext], 0, len(servers))

	for _, server := range servers {
		if httpParams, ok := server.StreamableHTTPParams(); ok {
			url := strings.TrimSpace(httpParams.URL)
			if url == "" {
				continue
			}
			authorization := strings.TrimSpace(httpParams.Authorization)
			params := llmmcp.NewMCPStreamableHTTPParams(url, authorization)
			toolkit := llmmcp.NewMCPToolkit(llmmcp.StaticMCPInit[*MyContext](params))
			toolkits = append(toolkits, toolkit)
			continue
		}

		if stdioParams, ok := server.StdioParams(); ok {
			command := strings.TrimSpace(stdioParams.Command)
			if command == "" {
				continue
			}
			args := make([]string, 0, len(stdioParams.Args))
			for _, arg := range stdioParams.Args {
				trimmed := strings.TrimSpace(arg)
				if trimmed != "" {
					args = append(args, trimmed)
				}
			}
			params := llmmcp.NewMCPStdioParams(command, args)
			toolkit := llmmcp.NewMCPToolkit(llmmcp.StaticMCPInit[*MyContext](params))
			toolkits = append(toolkits, toolkit)
		}
	}

	if len(toolkits) == 0 {
		return nil
	}

	return toolkits
}

func ptr[T any](v T) *T {
	return &v
}

File: agent-rust/examples/server/agent.rs
use chrono::Utc;
use llm_agent::{Agent, AgentTool};
use llm_sdk::{AudioOptions, LanguageModel, Modality, ReasoningOptions};
use std::sync::Arc;

use crate::{
    artifacts_tools::{
        ArtifactCreateTool, ArtifactDeleteTool, ArtifactGetTool, ArtifactListTool,
        ArtifactUpdateTool,
    },
    context::MyContext,
    finance_tools::{GetCryptoPriceTool, GetStockPriceTool},
    get_model::ModelInfo,
    information_tools::{GetNewsTool, SearchWikipediaTool},
    weather_tools::{GetCoordinatesTool, GetWeatherTool},
};

#[derive(Clone)]
pub struct AgentOptions {
    pub enabled_tools: Option<Vec<String>>,
    pub disabled_instructions: bool,
    pub temperature: Option<f64>,
    pub top_p: Option<f64>,
    pub top_k: Option<i32>,
    pub frequency_penalty: Option<f64>,
    pub presence_penalty: Option<f64>,
    pub audio: Option<AudioOptions>,
    pub reasoning: Option<ReasoningOptions>,
    pub modalities: Option<Vec<Modality>>,
}

pub fn get_available_tools() -> Vec<Box<dyn AgentTool<MyContext> + Send + Sync>> {
    vec![
        Box::new(ArtifactCreateTool),
        Box::new(ArtifactUpdateTool),
        Box::new(ArtifactGetTool),
        Box::new(ArtifactListTool),
        Box::new(ArtifactDeleteTool),
        Box::new(GetStockPriceTool),
        Box::new(GetCryptoPriceTool),
        Box::new(SearchWikipediaTool),
        Box::new(GetNewsTool),
        Box::new(GetCoordinatesTool),
        Box::new(GetWeatherTool),
    ]
}

pub fn create_agent(
    model: Arc<dyn LanguageModel + Send + Sync>,
    _model_info: &ModelInfo,
    options: &AgentOptions,
) -> Agent<MyContext> {
    let mut builder = llm_agent::AgentParams::new("MyAgent", model);

    if !options.disabled_instructions {
        builder = builder
            .add_instruction(
                "Answer in markdown format.\\nTo access certain tools, the user may have to \
                 provide corresponding API keys in the context fields on the UI.",
            )
            .add_instruction(|context: &MyContext| {
                let name = context.name.as_deref().unwrap_or("<not provided>");
                let location = context.location.as_deref().unwrap_or("<not provided>");
                let language = context.language.as_deref().unwrap_or("<not provided>");
                Ok(format!(
                    "The user name is {name}.\\nThe user location is {location}.\\nThe user \
                     speaks {language} language."
                ))
            })
            .add_instruction(|_context: &MyContext| {
                Ok(format!(
                    "The current date is {}.",
                    Utc::now().format("%a %b %d %Y")
                ))
            })
            .add_instruction(
                "For substantive deliverables (documents/specs/code), use the artifact tools (artifact_create, artifact_update, artifact_get, artifact_list, artifact_delete).\\nKeep chat replies brief and put the full document content into artifacts via these tools, rather than pasting large content into chat. Reference documents by their id.",
            );
    }

    // Add tools based on enabled_tools filter
    let enabled_tools = options.enabled_tools.as_ref();

    if enabled_tools.is_none()
        || enabled_tools
            .unwrap()
            .contains(&"get_stock_price".to_string())
    {
        builder = builder.add_tool(GetStockPriceTool);
    }
    if enabled_tools.is_none()
        || enabled_tools
            .unwrap()
            .contains(&"get_crypto_price".to_string())
    {
        builder = builder.add_tool(GetCryptoPriceTool);
    }
    if enabled_tools.is_none()
        || enabled_tools
            .unwrap()
            .contains(&"search_wikipedia".to_string())
    {
        builder = builder.add_tool(SearchWikipediaTool);
    }
    if enabled_tools.is_none() || enabled_tools.unwrap().contains(&"get_news".to_string()) {
        builder = builder.add_tool(GetNewsTool);
    }
    if enabled_tools.is_none()
        || enabled_tools
            .unwrap()
            .contains(&"get_coordinates".to_string())
    {
        builder = builder.add_tool(GetCoordinatesTool);
    }
    if enabled_tools.is_none() || enabled_tools.unwrap().contains(&"get_weather".to_string()) {
        builder = builder.add_tool(GetWeatherTool);
    }
    if enabled_tools.is_none()
        || enabled_tools
            .unwrap()
            .contains(&"artifact_create".to_string())
    {
        builder = builder.add_tool(ArtifactCreateTool);
    }
    if enabled_tools.is_none()
        || enabled_tools
            .unwrap()
            .contains(&"artifact_update".to_string())
    {
        builder = builder.add_tool(ArtifactUpdateTool);
    }
    if enabled_tools.is_none() || enabled_tools.unwrap().contains(&"artifact_get".to_string()) {
        builder = builder.add_tool(ArtifactGetTool);
    }
    if enabled_tools.is_none()
        || enabled_tools
            .unwrap()
            .contains(&"artifact_list".to_string())
    {
        builder = builder.add_tool(ArtifactListTool);
    }
    if enabled_tools.is_none()
        || enabled_tools
            .unwrap()
            .contains(&"artifact_delete".to_string())
    {
        builder = builder.add_tool(ArtifactDeleteTool);
    }

    builder = builder.max_turns(5);

    if let Some(temperature) = options.temperature {
        builder = builder.temperature(temperature);
    }
    if let Some(top_p) = options.top_p {
        builder = builder.top_p(top_p);
    }
    if let Some(top_k) = options.top_k {
        builder = builder.top_k(top_k);
    }
    if let Some(frequency_penalty) = options.frequency_penalty {
        builder = builder.frequency_penalty(frequency_penalty);
    }
    if let Some(presence_penalty) = options.presence_penalty {
        builder = builder.presence_penalty(presence_penalty);
    }

    if let Some(audio) = &options.audio {
        builder = builder.audio(audio.clone());
    }

    if let Some(reasoning) = &options.reasoning {
        builder = builder.reasoning(reasoning.clone());
    }

    if let Some(modalities) = &options.modalities {
        builder = builder.modalities(modalities.clone());
    }

    builder.build()
}
